azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -273,6 +273,109 @@ def _handle_pipeline_event(self, event):
                 logger.warning("incoming pipeline event with no handler.  dropping.")
 
 
+# NOTE: This stage could be a candidate for being refactored into some kind of other
+# pipeline-related structure. What's odd about it as a stage is that it doesn't really respond
+# to operations or events so much as it spawns them on a timer.
+# Perhaps some kind of... Pipeline Daemon?
+class SasTokenRenewalStage(PipelineStage):
+    # Amount of time, in seconds, prior to token expiration, when the renewal process will begin
+    DEFAULT_TOKEN_RENEWAL_MARGIN = 120
+
+    def __init__(self):
+        super(SasTokenRenewalStage, self).__init__()
+        self._token_renewal_timer = None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, op):
+        if (
+            isinstance(op, pipeline_ops_base.InitializePipelineOperation)
+            and self.pipeline_root.pipeline_configuration.sastoken
+        ):
+            self._start_renewal_timer()
+            self.send_op_down(op)
+        else:
+            self.send_op_down(op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _cancel_token_renewal_timer(self):
+        """Cancel and delete any pending renewal timer"""
+        timer = self._token_renewal_timer
+        self._token_renewal_timer = None
+        if timer:
+            logger.debug("Cancelling SAS Token renewal timer")
+            timer.cancel()
```

Comment:
>timer.cancel() [](start = 12, length = 14)

can you also put `timer = None`? For reasons that I don't quite remember i see this pattern in all timers in our stack

Reply:
This happens on line 303

---
In reply to: [426955668](https://github.com/Azure/azure-iot-sdk-python/pull/555#discussion_r426955668) [](ancestors = 426955668)